### PR TITLE
Fix toolbar buttons being blocked by its own overlays

### LINKED ISSUE #2765: Only master volume can be changed while something else (e.g. notifications) has focus
While using the scroll wheel Master volume will change. When you hover over Effects or Music then that volume can be changed. But only when something else isn't opened, like notifications or music player.

### LINKED ISSUE #2898: While the notifications tab is open, the volume overlay still fades when hovered over

### PR DESCRIPTION
**Pull request by Joehuu**
- ~Replaces `ToolbarUserArea`'s `Action` with `StateContainer`~
- ~Removes specified depths~
- ~Sorts overlays by depth in code (what goes top/bottom)~
- ~Fixes offset of attaching overlays if not already~
- ~Moves volume overlay above other overlays excluding settings (fixes #2765 & fixes #2898)~

~Edit: Not sure about the behavior on `LoginOverlay`, `MusicController`, `NotificationOverlay` as they overlap and look bad.~ ~Latest commit tries to fix this.~ Still an issue as music controller is not a `singleDisplaySideOverlay`.

*Edit: My bad if this PR went off the scope of the initial fix.*

---

### CODE REVIEW

**Comment by peppy** [comment]
Might be worth checking if we even need to specify the depths. Since they are loaded using single file logic now, their load order may be guaranteed enough to remove (as long as they aren't interacting with components outside the single file queue or ordered non sequentially)

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -490,12 +480,13 @@ protected override void LoadComplete()
...
             dependencies.Cache(dialogOverlay);
             dependencies.Cache(accountCreation);
+            dependencies.Cache(loginOverlay);
 
             chatOverlay.StateChanged += state => channelManager.HighPollRate.Value = state == Visibility.Visible;
 
             Add(externalLinkOpener = new ExternalLinkOpener());
 
-            var singleDisplaySideOverlays = new OverlayContainer[] { settings, notifications };
+            var singleDisplaySideOverlays = new OverlayContainer[] { settings, notifications, musicController, loginOverlay };
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: music controller shouldn't be in here

Comment (Joehuu) [reply osu.Game/OsuGame.cs]: Is there a reason why?

**Comment by Joehuu** [comment]
Not sure if I should push https://github.com/Joehuu/osu/compare/fix-toolbar-buttons...Joehuu:toolbar-buttons-rework. It removes the individual toolbar offset of some overlays as most overlays need that offset already.

File: osu.Game/OsuGame.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/OsuGame.cs]: it's supposed to stay visible even when other overlays are opened/closed. at this point you've added every overlay inside `singleDisplay` which is defeating the purpose of it existing in the first place

**Comment by peppy** [comment]
I think it can be included in this PR.

File: osu.Game/OsuGame.cs [hunk 2]
```diff
@@ -414,68 +413,56 @@ protected override void LoadComplete()
...
-            loadComponentSingleFile(musicController = new MusicController
-            {
-                Depth = -5,
-                Position = new Vector2(0, Toolbar.HEIGHT),
-                Anchor = Anchor.TopRight,
-                Origin = Anchor.TopRight,
-            }, floatingOverlayContent.Add);
 
+            //regular overlay elements
+            loadComponentSingleFile(direct = new DirectOverlay(), overlayContent.Add);
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: you've changed the order of loading here. is this intentional/tested?

File: osu.Game/OsuGame.cs [hunk 3]
```diff
@@ -490,12 +477,13 @@ protected override void LoadComplete()
...
             dependencies.Cache(dialogOverlay);
             dependencies.Cache(accountCreation);
+            dependencies.Cache(loginOverlay);
 
             chatOverlay.StateChanged += state => channelManager.HighPollRate.Value = state == Visibility.Visible;
 
             Add(externalLinkOpener = new ExternalLinkOpener());
 
-            var singleDisplaySideOverlays = new OverlayContainer[] { settings, notifications };
+            var singleDisplaySideOverlays = new OverlayContainer[] { settings, notifications, loginOverlay };
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: this means that if you trigger a login popup from setting, the settings are going to disappear.  not sure this is what we want.

Comment (Joehuu) [reply osu.Game/OsuGame.cs]: If login and settings overlays are both open, the bottom overlay would be blocked (settings) with the current behavior of overlays.

With the dimming behind both of these overlays, both seem clickable (but the bottom overlay is blocked).

**Comment by Joehuu** [comment]
Will split this PR for faster merging.

**Pull request closed without merging**